design/proposal-crossplane-v2.md

```diff
@@ -0,0 +1,948 @@
+
+The application isn’t managed by Crossplane. Instead it claims (uses)
+infrastructure that’s defined and managed by Crossplane. This is an artifact of
+Crossplane’s early design. We took the Kubernetes claim model used by
+[persistent volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/)
+and extended it to handle any kind of infrastructure.
+
+I believe two things make Crossplane awkward for composing applications:
+
+1. Crossplane XRs and MRs are cluster scoped, while most Kubernetes resources
+   are namespaced
+2. Crossplane [doesn’t support](https://github.com/crossplane/crossplane/issues/1730)
+   directly composing arbitrary (non-Crossplane) Kubernetes resources
+
+When you build a Crossplane XR that represents an application today, it looks
+like this:
+
+```mermaid
+flowchart LR
+  subgraph ns [my-namespace]
+    app["Application (Claim)"]
+    dply["Deployment"]
+    svc["Service"]
+  end
+
+  %% Cluster scoped stuff
+  xapp["XApplication (XR)"]
+  obj1["Object (MR)"]
+  obj2["Object (MR)"]
+  rds["RDSInstance (MR)"]
+
+  app --> xapp
+  xapp --> obj1
+  xapp --> obj2
+  xapp --> rds
+  obj1 --> dply
+  obj2 --> svc
+```
+
+The application developer creates an Application in their namespace. This is a
+namespaced claim. Creating the claim ultimately causes Crossplane to create a
+namespaced Deployment and Service, and a cluster scoped RDSInstance. These are
+the composed resources. Between the claim and composed resources there’s a
+cluster scoped XR. For the Deployment and Service there’s also cluster scoped
+provider-kubernetes Object MRs.
+
+What value does the creator of the Application get from the intermediary cluster
+scoped resources? Why is the RDSInstance MR cluster scoped while the Deployment
+and Service aren’t? They’re all components of the same application.
+
+The following architecture achieves the same goal, but is simpler and more
+intuitive:
+
+```mermaid
+flowchart LR
+  subgraph ns [my-namespace]
+    direction LR
+    xr["Application (XR)"]
+  	dply["Deployment"]
+    svc["Service"]
+    rds["RDSInstance (MR)"]
+  end
+
+  xr --> dply
+  xr --> svc
+  xr --> rds
+```
+
+The namespaced Application abstraction is directly composed of the three
+resources that make up the logical application - a Deployment, a Service, and an
+RDSInstance. They’re all composed in the same namespace as the abstraction.
+
+This document proposes a set of changes to Crossplane that enable this simpler
+architecture.
+
+## Goals
+
+The goal of this proposal is to:
+
+* Make Crossplane less opinionated and more broadly useful
+* Make Crossplane more intuitive and easier to use
+* Avoid breaking existing Crossplane users
+
+These goals are in service of making Crossplane a better tool for composing
+applications _and_ infrastructure.
+
+## Proposal
+
+I propose we make the following changes in Crossplane v2.
+
+In core I propose we:
```

Comment:
Do we believe that in V1 we should natively support asynchronous reconciliation of resources or do we feel that is more of a V2 feature?